internal/liveness/metrics.go

```diff
@@ -0,0 +1,252 @@
+		secretNamespace = pvInfo.Spec.CSI.ControllerExpandSecretRef.Namespace
+	} else {
+		secretName = pvInfo.Spec.CSI.NodePublishSecretRef.Name
+		secretNamespace = pvInfo.Spec.CSI.NodePublishSecretRef.Namespace
+	}
+	imageCluster := pvInfo.Spec.CSI.VolumeAttributes["clusterID"]
+	monitors, err := util.Mons(csiConfigFile, imageCluster)
+	if err != nil {
+		klog.Errorf(err.Error())
+		return
+	}
+	var cr *util.Credentials
+	if value, exist := secrets[secretNamespace+"/"+secretName]; exist {
+		cr = value
+	} else {
+		var secret *v1.Secret
+		secret, err = client.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		newSecrets := make(map[string]string)
+		for key, value := range secret.Data {
+			newSecrets[key] = string(value)
+		}
+		if _, ok := newSecrets["adminID"]; ok {
+			cr, err = util.NewAdminCredentials(newSecrets)
+		} else {
+			cr, err = util.NewUserCredentials(newSecrets)
+		}
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		secrets[secretNamespace+"/"+secretName] = cr
+	}
+	imagePool := pvInfo.Spec.CSI.VolumeAttributes["pool"]
+	volumeName := pvInfo.Spec.CSI.VolumeAttributes["fsName"]
+	switch {
+	case imagePool == "" && volumeName == "":
+		// unknown volume
+		klog.Warningf("Found unknown type pvc %s with pvName %s", pvcName, pvName)
+		return
+	case volumeName != "":
+		// cephfs volume
+		subvolumeGroup, err := util.CephFSSubvolumeGroup(csiConfigFile, imageCluster)
+		if err != nil {
+			klog.Errorf("Failed to get subvolumeGroup: %s", err.Error())
+			return
+		}
+		if cephfs.CheckSubVolumeInCluster(monitors, imageName, subvolumeGroup, volumeName, cr) {
+			cephfsMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, volumeName, subvolumeGroup, imageName).Set(1)
+		} else {
+			cephfsMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, volumeName, subvolumeGroup, imageName).Set(0)
+		}
+	default:
+		// rbd volume
+		err := rbd.GetImageInfo(context.TODO(), monitors, cr, imagePool, imageName)
+		if err != nil {
+			rbdMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, imagePool, imageName).Set(0)
+			util.ExtendedLogMsg(err.Error())
+		} else {
+			rbdMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, imagePool, imageName).Set(1)
+		}
+	}
+}
+
+func getPvcMappings(ctx context.Context, client *k8s.Clientset, pvcs <-chan v1.PersistentVolumeClaim, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case pvc, ok := <-pvcs:
+			if !ok {
+				return
+			}
+			collectPvcMapping(client, &pvc, secrets)
+		case <-ctx.Done():
+			klog.Errorf("Collect mapping metrics timeout , you can increase scrapeMetricsGoroutines or scrapeMetricsTimeout")
+			return
+		}
+	}
+}
+
+func recordMapping(pollTime, pollTimeout time.Duration, csiConn *grpc.ClientConn, scapeTimeout time.Duration, numGoroutines int) {
+	// register prometheus metrics
+	vType := os.Getenv("DRIVER_TYPE")
```

Comment:
I don't think this works as intended, where did you get `DRIVER_TYPE` from?

Reply:
I add the ENV variable to deploy template to distinguish between csi-rbd and csi-cephfs.